src/storage.ts

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2021 Google Inc. All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+import { openDB, DBSchema, IDBPDatabase } from 'idb';
+
+const mainDBName = 'appDB';
+const mainImageName = 'mainImage';
+
+export type FileUniqueID = string;  // generated by the storage layer
+export type Timestamp = number;
+
+export type FileMetadata = {
+  filename: string;
+  mimeType: string;
+  title?: string;
+}
+
+export type FileRecord = {
+  id: FileUniqueID;
+  metadata: FileMetadata;
+  added: Timestamp;
+  blobModified: Timestamp;
+  metadataModified: Timestamp;
+};
+
+interface LlamaDB extends DBSchema {
+  blob: {
+    key: string;
+    value: Blob;
+  };
+  metadata: {
+    key: string;
+    value: FileRecord;
+  };
+}
+
+export default class LlamaStorage {
```

Comment:
Style item for discussion: I dislike default exports, and would prefer that we export everything by name. In this particular case, `storage.ts` exports four types and one class, none of which are called `storage`.

It's not super significant, the difference between:

```typescript
import LlamaStorage, { type FileRecord } from './storage.ts';
import Storage, { type FileRecord } from './storage.ts';  // implicitly identical

import { LlamaStorage, type FileRecord } from './storage.ts';
import { Storage, type FileRecord } from './storage.ts';  // error
```

Reply:
done.